p2panda-net/src/sync/initiate.rs

```diff
@@ -0,0 +1,173 @@
+                                .into(),
+                        ));
+                    }
+                    sync_handshake_success = true;
+
+                    // Inform the engine that we are expecting sync messages from the peer on this
+                    // topic.
+                    engine_actor_tx
+                        .send(ToEngineActor::SyncHandshakeSuccess {
+                            peer,
+                            topic: topic.clone(),
+                        })
+                        .await
+                        .map_err(|err| {
+                            SyncError::Critical(format!(
+                                "engine_actor_tx failed sending sync handshake success: {err}"
+                            ))
+                        })?;
+
+                    continue;
+                }
+
+                // 2. Data Sync Phase.
+                // ~~~~~~~~~~~~~~~~~~~
+                let FromSync::Data(header, payload) = message else {
+                    return Err(SyncError::Critical("expected to receive only data messages from sync session in data sync phase".into()));
+                };
+
+                engine_actor_tx
+                    .send(ToEngineActor::SyncMessage {
+                        header,
+                        payload,
+                        delivered_from: peer,
+                        topic: topic.clone(),
+                    })
+                    .await
+                    .map_err(|err| {
+                        SyncError::Critical(format!(
+                            "engine_actor_tx failed sending sync message: {err}"
+                        ))
+                    })?;
+            }
+
+            Ok(())
+        })
+    };
+
+    // Run the "initiating peer" side of the sync protocol.
+    let result = sync_protocol
+        .initiate(
+            topic.clone(),
+            Box::new(&mut send),
+            Box::new(&mut recv),
+            Box::new(&mut sink),
+        )
+        .await;
+
+    // Drop the tx, so the rx in the glue task receives the closing event.
+    drop(sink);
+
+    // We're expecting the task to exit with a result soon, we're awaiting it here ..
+    let glue_task_result = glue_task_handle
+        .await
+        .map_err(|err| SyncError::Critical(format!("glue task handle failed: {err}")))?;
+
+    // .. to inform some brave developer who will read the error logs ..
+    if let Err(SyncError::Critical(err)) = &glue_task_result {
+        error!("critical error in sync protocol: {err}");
+    }
+
+    // .. and forward it further.
+    glue_task_result?;
+
+    // The same we're doing with errors coming from the sync protocol implementation itself.
+    if let Err(err) = result {
+        match &err {
+            SyncError::Critical(err) => {
+                error!("critical error in sync protocol: {err}");
+            }
+            _ => {
+                warn!("error in sync protocol: {err}");
+            }
+        }
```

Comment:
Similarly to above, I'm not sure if this is not too much logging (we might have other places which already log on sync errors).

Reply:
In this case we also log in `src/sync/manager.rs` L277 or L291 (depending on error type) but we don't differentiate between sync error types (ie. we don't specifically match on a critical error).

Again, would be nice to collapse these into a single log (just because it might be confusing when debugging in the future: "is this one thing going wrong or two?").

Reply:
Okay, let me remove one of them then 👍🏻